File: testbed_ruby/test/associations/m_n_to_optional_n_test.rb
require 'test/unit'
require './src-gen-umple/student_a_b'
require './src-gen-umple/mentor_a_b'
require './src-gen-umple/program_a_b'

module CruiseAssociations
class MNToOptionalNTest < Test::Unit::TestCase

  def test_CreateStudentWihtoutMentor
  
    s = StudentAB.new(99)
    assert_equal(false,s.is_number_of_mentors_valid)
  end
  
  def test_CreateMentorWithoutStudent
    m = MentorAB.new("blah")
    assert_equal(0,m.number_of_students)
  end
  
  def test_AddStudentsAndMentorsOkay
    m = MentorAB.new("blah")
    m2 = MentorAB.new("blah")
    m3 = MentorAB.new("blah")
    m4 = MentorAB.new("blah")
    m5 = MentorAB.new("blah")
    m6 = MentorAB.new("blah")
    
    s = StudentAB.new(99)
    s2 = StudentAB.new(98)
    s3 = StudentAB.new(97)

    assert_equal(true,m.add_student(s))
    assert_equal(true,m.add_student(s2))
    assert_equal(false,m.add_student(s3))
    
    assert_equal(true,s.add_mentor(m2))
    assert_equal(true,s.add_mentor(m3))
    assert_equal(true,s.add_mentor(m4))
    assert_equal(true,s.add_mentor(m5))
    assert_equal(false,s.add_mentor(m6))
    
    assert_equal(2, m.number_of_students)
    assert_equal(m4,s.get_mentor(3))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(5,s.number_of_mentors)
  end   
  
  def test_RemoveMiddleMentorWhenNotValidMaintainsTheOrder
    s = StudentAB.new(99)
    
    m = MentorAB.new("blah")
    m2 = MentorAB.new("blah")
    m3 = MentorAB.new("blah")

    s.add_mentor(m)
    s.add_mentor(m2)
    s.add_mentor(m3)
    assert_equal(false,s.remove_mentor(m2))
    assert_equal(3, s.number_of_mentors)
    assert_equal(m2,s.get_mentor(1))
  end 
  
  def test_RemoveStudents
    m = MentorAB.new("blah")
    m2 = MentorAB.new("blah2")
    m3 = MentorAB.new("blah3")
    m4 = MentorAB.new("blah4")
    
    s = StudentAB.new(99)
    s2 = StudentAB.new(98)
    s3 = StudentAB.new(97)
    
    m.add_student(s)
    m.add_student(s2)
    m2.add_student(s)
    m2.add_student(s2)
    m3.add_student(s)
    m3.add_student(s2)
    m4.add_student(s)
    m4.add_student(s2)
    
    assert_equal(false,m.remove_student(s3))
    assert_equal(true,m.remove_student(s2))
    
    assert_equal(3,s2.number_of_mentors)
    assert_equal(1, m.number_of_students)
  end 
  
  def test_SetStudentsTooMany
    m = MentorAB.new("blah")
    
    s = StudentAB.new(99)
    s2 = StudentAB.new(98)
    s3 = StudentAB.new(97)
    
    assert_equal(true,m.add_student(s))
    assert_equal(true,m.add_student(s2))
    assert_equal(false,m.add_student(s3))
    
    assert_equal(2, m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(0,s3.number_of_mentors)
  end   
  
  def test_MentorAlreadyHasEnoughStudents
    m = MentorAB.new("blah")
    
    s = StudentAB.new(99)
    s2 = StudentAB.new(98)
    s3 = StudentAB.new(97)
    
    m.add_student(s)
    m.add_student(s2)
    
    assert_equal(false,m.add_student(s3))
    
    assert_equal(2, m.number_of_students)
    assert_equal(0,s3.number_of_mentors)
  end    
  

  def test_getBoundsForStudent
    assert_equal(2,MentorAB::maximum_number_of_students)
  end  
  
  def test_getBoundsForMentor
    assert_equal(3,StudentAB::minimum_number_of_mentors)
    assert_equal(5,StudentAB::maximum_number_of_mentors)
  end    

  def test_deleteMentorAndStudentHasEnough
    m = MentorAB.new("blah")
    m2 = MentorAB.new("blah2")
    m3 = MentorAB.new("blah2")
    m4 = MentorAB.new("blah2")

    s = StudentAB.new(99)
    s2 = StudentAB.new(98)
    
    m.add_student(s)
    m.add_student(s2)
    m2.add_student(s)
    m2.add_student(s2)
    m3.add_student(s)
    m3.add_student(s2)
    m4.add_student(s)
    m4.add_student(s2)
    
    p = ProgramAB.new
    s.set_program(p)
    
    m.delete
    assert_equal(0,m.number_of_students)
    assert_equal(p,s.get_program)
    assert_equal(3,s.number_of_mentors)
  end  

  def test_deleteMentorAndStudentNowHasTooFewMentors
    m = MentorAB.new("blah")
    m2 = MentorAB.new("blah2")
    m3 = MentorAB.new("blah3")

    s = StudentAB.new(99)
    s2 = StudentAB.new(98)
    
    m.add_student(s)
    m.add_student(s2)
    m2.add_student(s)
    m2.add_student(s2)
    m3.add_student(s)
    m3.add_student(s2)
    
    p = ProgramAB.new
    s.set_program(p)

    assert_equal(2,m.number_of_students)
    assert_equal(2,m2.number_of_students)
    assert_equal(2,m3.number_of_students)
    assert_equal(3,s.number_of_mentors)
    assert_equal(3,s2.number_of_mentors)
    
    m.delete
    assert_equal(0,m.number_of_students)
    assert_equal(0,m2.number_of_students)
    assert_equal(0,m3.number_of_students)
    assert_equal(0,s.number_of_mentors)
    assert_equal(0,s2.number_of_mentors)
  end  

end
end

File: testbed_php/test/associations/MNToOptionalNTest.php
<?php

class MNToOptionalNTest extends UnitTestCase
{


  public function xtest_CreateStudentWihtoutMentor()
  {
    $s = new StudentAB(99);
    $this->assertEqual(false,$s->isNumberOfMentorsValid());
  }
  

  public function xtest_CreateMentorWithoutStudent()
  {
    $m = new MentorAB("blah");
    $this->assertEqual(0,$m->numberOfStudents());
  }
  

  public function xtest_AddStudentsAndMentorsOkay()
  {
    $m = new MentorAB("blah");
    $m2 = new MentorAB("blah");
    $m3 = new MentorAB("blah");
    $m4 = new MentorAB("blah");
    $m5 = new MentorAB("blah");
    $m6 = new MentorAB("blah");
    
    $s = new StudentAB(99);
    $s2 = new StudentAB(98);
    $s3 = new StudentAB(97);

    $this->assertEqual(true,$m->addStudent($s));
    $this->assertEqual(true,$m->addStudent($s2));
    $this->assertEqual(false,$m->addStudent($s3));
    
    $this->assertEqual(true,$s->addMentor($m2));
    $this->assertEqual(true,$s->addMentor($m3));
    $this->assertEqual(true,$s->addMentor($m4));
    $this->assertEqual(true,$s->addMentor($m5));
    $this->assertEqual(false,$s->addMentor($m6));
    
    $this->assertEqual(2, $m->numberOfStudents());
    $this->assertEqual($m4,$s->getMentor_index(3));
    $this->assertEqual($m,$s2->getMentor_index(0));
    $this->assertEqual(5,$s->numberOfMentors());
  }   
  

  public function xtest_RemoveMiddleMentorWhenNotValidMaintainsTheOrder()
  {
    $s = new StudentAB(99);
    
    $m = new MentorAB("blah");
    $m2 = new MentorAB("blah");
    $m3 = new MentorAB("blah");

    $s->addMentor($m);
    $s->addMentor($m2);
    $s->addMentor($m3);
    $this->assertEqual(false,$s->removeMentor($m2));
    $this->assertEqual(3, $s->numberOfMentors());
    $this->assertEqual($m2,$s->getMentor_index(1));
  } 
  

  public function xtest_RemoveStudents()
  {
    $m = new MentorAB("blah");
    $m2 = new MentorAB("blah2");
    $m3 = new MentorAB("blah3");
    $m4 = new MentorAB("blah4");
    
    $s = new StudentAB(99);
    $s2 = new StudentAB(98);
    $s3 = new StudentAB(97);
    
    $m->addStudent($s);
    $m->addStudent($s2);
    $m2->addStudent($s);
    $m2->addStudent($s2);
    $m3->addStudent($s);
    $m3->addStudent($s2);
    $m4->addStudent($s);
    $m4->addStudent($s2);
    
    $this->assertEqual(false,$m->removeStudent($s3));
    $this->assertEqual(true,$m->removeStudent($s2));
    
    $this->assertEqual(3,$s2->numberOfMentors());
    $this->assertEqual(1, $m->numberOfStudents());
  } 
  

  public function xtest_SetStudentsTooMany()
  {
    $m = new MentorAB("blah");
    
    $s = new StudentAB(99);
    $s2 = new StudentAB(98);
    $s3 = new StudentAB(97);
    
    $this->assertEqual(true,$m->addStudent($s));
    $this->assertEqual(true,$m->addStudent($s2));
    $this->assertEqual(false,$m->addStudent($s3));
    
    $this->assertEqual(2, $m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor_index(0));
    $this->assertEqual($m,$s2->getMentor_index(0));
    $this->assertEqual(0,$s3->numberOfMentors());
  }   
  

  public function xtest_MentorAlreadyHasEnoughStudents()
  {
    $m = new MentorAB("blah");
    
    $s = new StudentAB(99);
    $s2 = new StudentAB(98);
    $s3 = new StudentAB(97);
    
    $m->addStudent($s);
    $m->addStudent($s2);
    
    $this->assertEqual(false,$m->addStudent($s3));
    
    $this->assertEqual(2, $m->numberOfStudents());
    $this->assertEqual(0,$s3->numberOfMentors());
  }    
  

  public function xtest_getBoundsForStudent()
  {
    $this->assertEqual(2,MentorAB::maximumNumberOfStudents());
  }  
  

  public function xtest_getBoundsForMentor()
  {
    $this->assertEqual(3,StudentAB::minimumNumberOfMentors());
    $this->assertEqual(5,StudentAB::maximumNumberOfMentors());
  }    
  

  public function xtest_deleteMentorAndStudentHasEnough()
  {
    
    $m = new MentorAB("blah");
    $m2 = new MentorAB("blah2");
    $m3 = new MentorAB("blah2");
    $m4 = new MentorAB("blah2");

    $s = new StudentAB(99);
    $s2 = new StudentAB(98);
    
    $m->addStudent($s);
    $m->addStudent($s2);
    $m2->addStudent($s);
    $m2->addStudent($s2);
    $m3->addStudent($s);
    $m3->addStudent($s2);
    $m4->addStudent($s);
    $m4->addStudent($s2);
    
    $p = new ProgramAB();
    $s->setProgram($p);
    
    $m->delete();
    $this->assertEqual(0,$m->numberOfStudents());
    $this->assertEqual($p,$s->getProgram());
    $this->assertEqual(3,$s->numberOfMentors());
  }  


  public function test_deleteMentorAndStudentNowHasTooFewMentors()
  {
    
    $m = new MentorAB("blah");
    $m2 = new MentorAB("blah2");
    $m3 = new MentorAB("blah3");

    $s = new StudentAB(99);
    $s2 = new StudentAB(98);
    
    $m->addStudent($s);
    $m->addStudent($s2);
    $m2->addStudent($s);
    $m2->addStudent($s2);
    $m3->addStudent($s);
    $m3->addStudent($s2);
    
    $p = new ProgramAB();
    $s->setProgram($p);

    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual(2,$m2->numberOfStudents());
    $this->assertEqual(2,$m3->numberOfStudents());
    $this->assertEqual(3,$s->numberOfMentors());
    $this->assertEqual(3,$s2->numberOfMentors());
    
    $m->delete();
    $this->assertEqual(0,$m->numberOfStudents());
    $this->assertEqual(0,$m2->numberOfStudents());
    $this->assertEqual(0,$m3->numberOfStudents());
    $this->assertEqual(0,$s->numberOfMentors());
    $this->assertEqual(0,$s2->numberOfMentors());
  }  

}
